scripts/zones/RoMaeve/npcs/Goblin_Footprint.lua

```diff
@@ -0,0 +1,66 @@
+-----------------------------------
+-- Area: Bostaunieux Oubliette
+--  NPC: Goblin Footprint
+-- Type: NPC
+-- !pos  111.806 -24.007 57.266 167
+-----------------------------------
+local ID = require("scripts/zones/RoMaeve/IDs")
+require("scripts/globals/pasteventwatcher")
+require("scripts/globals/quests")
+require("scripts/globals/missions")
+-----------------------------------
+---
+function onTrade(player,npc,trade)
+    local options = {}
+    if player:hasCompletedMission(WINDURST,dsp.mission.id.windurst.VAIN) == true then
+        options[1] = 1
+    end
+    if player:hasCompletedMission(WINDURST,dsp.mission.id.windurst.MOON_READING) == true then
+        options[2] = 1
+    end
+    if player:getQuestStatus(ZILART,dsp.quest.id.outlands.DIVINE_MIGHT) == QUEST_COMPLETED then
+        options[3] = 1
+    end
+    if player:hasCompletedMission(ASA,dsp.mission.id.asa.FOUNTAIN_OF_TROUBLE) == true then
+        options[4] = 1
+    end
+    local eventList = dsp.pasteventwatcher.goblinFootprint({options[1], options[2], options[3], options[4]})
+    if not next(options) == nil then
+        player:startEvent(6, eventList, 4294967294, 4294967294, 4294967294)
+    end
+end
+
+function onTrigger(player,npc)
+    if player:hasCompletedMission(WINDURST,dsp.mission.id.windurst.VAIN) == true or
+        player:hasCompletedMission(WINDURST,dsp.mission.id.windurst.MOON_READING) == true or
+        player:getQuestStatus(ZILART,dsp.quest.id.outlands.DIVINE_MIGHT) == QUEST_COMPLETED or
+        player:hasCompletedMission(ASA,dsp.mission.id.asa.FOUNTAIN_OF_TROUBLE) == true then
```

Comment:
It'd be nice to have a helper function that accepts a player and table of quests/missions, and then depending on a passed in flag, either builds the full options table (above) for the trade, or simply goes through the checks and returns true on the first successful check (for checking if the goblin should ask for items).

I don't know if it'd be best to come up with a way of describing quest/mission checks in a table, so a table could be defined for each specific goblin:
```
local checks =
{
    {mission, windurst, vain},
    {mission, windurst, moon_reading}
}
```
And then have the helper function build options tables or onTrigger check as appropriate; or

Just use a table of lambdas (kind of ugly, but easier to implement):
```
local checks =
{
    function(player) player:hasCompletedMission(WINDURST,dsp.mission.id.windurst.VAIN) end,
    function(player) player:hasCompletedMission(WINDURST,dsp.mission.id.windurst.MOON_READING) end
}
```

And then in the helper function, just invoke each of the lambdas in order (`checks[i](player)`) and store the result into a table. If the "on trigger checking" flag is passed into the helper, abort on the first invokation that's `true`, otherwise fill out the table and return that.

Reply:
you would break out of the loop if it returns false.

Reply:
Can make an assignment to the result (`blah = checks[i](player)`), and then check the assigned var to determine if you should keep going through your loop (`while ((not blah) or build_options) and i <= #checks`), rather than invoking each check directly and hitting a return that breaks you out.

(And if you should keep going, pop the result into the table for when the function is called with the "build options" flag)

```
function checkOptions(player, checks, build_options)
    local options = {}
    local check = false
    local i = 1
    while ((not check) or build_options and i <= #checks ) do
        check = checks[i](player)
        if check and not build_options then
            return true
        else
            options[i] = check
        end
    end
    if build_options then
        return options
    else
        return false
    end
end

local checks = {
    -- I'll spare the lambdas
}

onTrigger
    if not checkOptions(player, checks, false) then
        player:startEvent(6) -- You see nothing
    end
end

onTrade
    local options = checkOptions(player, checks, true)
end
```

Alternatively, if you're willing to sacrifice being able to abort early on the first true result for your onTrigger if check, you could just always build the table up front (without any sort of flag) and use a dual return.

```
function checkOptions(player, checks)
    local options = {}
    local has_options = false
    local i = 1
    while (i <= #checks ) do
        options[i] = checks[i](player)
        if options[i] then
            has_options = true
        end
    end
    return has_options, options
end

onTrigger
    local has_options, options = checkOptions(player, checks)
    if has_options then
        -- Items la la
    end
end

onTrade
    local has_options, options = checkOptions(player, checks)
    if has_options then
        player:startEvent(X, options)
    end
end
```